src/main/anovos/data_transformer/transformers.py

```diff
@@ -3430,39 +3431,46 @@ def outlier_categories(
         for index, i in enumerate(list_of_cols):
             window = Window.partitionBy().orderBy(F.desc("count_pct"))
             df_cats = (
-                idf.groupBy(i)
-                .count()
-                .dropna()
-                .withColumn(
-                    "count_pct",
-                    F.col("count") / F.sum("count").over(Window.partitionBy()),
-                )
-                .withColumn("rank", F.rank().over(window))
-                .withColumn(
-                    "cumu",
-                    F.sum("count_pct").over(
-                        window.rowsBetween(Window.unboundedPreceding, 0)
-                    ),
-                )
-                .withColumn("lag_cumu", F.lag("cumu").over(window))
-                .fillna(0)
-                .where(~((F.col("cumu") >= coverage) & (F.col("lag_cumu") >= coverage)))
-                .where(F.col("rank") <= (max_category - 1))
-                .select(F.lit(i).alias("attribute"), F.col(i).alias("parameters"))
+                idf.select(i).groupBy(i)
+                    .count()
+                    .dropna()
+                    .withColumn(
+                        "count_pct",
+                        F.col("count") / F.sum("count").over(Window.partitionBy()),
+                    )
+                    .withColumn("rank", F.rank().over(window))
+                    .withColumn(
+                        "cumu",
+                        F.sum("count_pct").over(
+                            window.rowsBetween(Window.unboundedPreceding, 0)
+                        ),
+                    )
+                    .withColumn("lag_cumu", F.lag("cumu").over(window))
+                    .fillna(0)
+                    .where(~((F.col("cumu") >= coverage) & (F.col("lag_cumu") >= coverage)))
+                    .where(F.col("rank") <= (max_category - 1))
+                    .select(F.lit(i).alias("attribute"), F.col(i).alias("parameters"))
             )
             if index == 0:
                 df_model = df_cats
             else:
                 df_model = df_model.union(df_cats)
 
+    df_params = df_model.rdd.groupByKey().mapValues(list).collect()
```

Comment:
It is the best if we can minimize the number of action (`collect`) in the code, optimally it should not be inside the function. Can you think of a way to only use transformation operation (`select`, `withColumn`, etc) to do this, without having to use `collect`.

Reply:
Here I minimised use of collect() to only once, earlier it was being used per column. 
What do you mean by this "optimally it should not be inside the function" ?
Regarding this "Can you think of a way to only use transformation operation (select, withColumn, etc) to do this, without having to use collect", I will have to work more on this and come back.